fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java

```diff
@@ -601,6 +601,63 @@ public void replayDropDb(String dbName, boolean isForceDrop, Long recycleTime) t
         }
     }
 
+    public void recoverDatabase(String dbName, long dbId, String newDbName) throws DdlException {
+        // check is new db with same name already exist
+        if (!Strings.isNullOrEmpty(newDbName)) {
+            if (getDb(newDbName).isPresent()) {
+                throw new DdlException("Database[" + newDbName + "] already exist.");
+            }
+        } else {
+            if (getDb(dbName).isPresent()) {
+                throw new DdlException("Database[" + dbName + "] already exist.");
+            }
+        }
+
+        Database db = Env.getCurrentRecycleBin().recoverDatabase(dbName, dbId);
+
+        // add db to catalog
+        if (!tryLock(false)) {
+            throw new DdlException("Failed to acquire catalog lock. Try again");
+        }
+        db.writeLock();
+        List<Table> tableList = db.getTablesOnIdOrder();
+        MetaLockUtils.writeLockTables(tableList);
+        try {
+            if (!Strings.isNullOrEmpty(newDbName)) {
+                if (fullNameToDb.containsKey(newDbName)) {
+                    throw new DdlException("Database[" + newDbName + "] already exist.");
+                    // it's ok that we do not put db back to CatalogRecycleBin
+                    // cause this db cannot recover any more
+                }
+            } else {
+                if (fullNameToDb.containsKey(db.getFullName())) {
+                    throw new DdlException("Database[" + db.getFullName() + "] already exist.");
+                    // it's ok that we do not put db back to CatalogRecycleBin
+                    // cause this db cannot recover any more
+                }
+            }
+            if (!Strings.isNullOrEmpty(newDbName)) {
+                try {
+                    db.writeUnlock();
+                    db.setNameWithLock(newDbName);
+                } finally {
+                    db.writeLock();
+                }
+            }
+            fullNameToDb.put(db.getFullName(), db);
+            idToDb.put(db.getId(), db);
+            // log
+            RecoverInfo recoverInfo = new RecoverInfo(db.getId(), -1L, -1L, newDbName, "", "");
+            Env.getCurrentEnv().getEditLog().logRecoverDb(recoverInfo);
+            db.unmarkDropped();
+        } finally {
+            MetaLockUtils.writeUnlockTables(tableList);
+            db.writeUnlock();
+            unlock();
+        }
+        LOG.info("recover database[{}]", db.getId());
+    }
+
     public void recoverDatabase(RecoverDbStmt recoverStmt) throws DdlException {
         // check is new db with same name already exist
         String newDbName = recoverStmt.getNewDbName();
```

Comment:
you can call the new recoverDatabase(String dbName, long dbId, String newDbName) function to remove duplicate code

Reply:
thanks have handled it